File: ObjectiveCDemo/TableViewControllerWithSections/TableViewControllerWithSections.m
#import "TableViewControllerWithSections.h"

@import DATAStack;
@import CoreData;

#import "ObjectiveCDemo-Swift.h"


static NSString *CellIdentifier = @"CellIdentifier";

@interface TableViewControllerWithSections ()

@property (nonatomic, weak) DATAStack *dataStack;
@property (nonatomic) DATASource *dataSource;

@end

@implementation TableViewControllerWithSections

- (instancetype)initWithDataStack:(DATAStack *)dataStack {
    self = [super initWithStyle:UITableViewStylePlain];
    if (self) {
        _dataStack = dataStack;
    }

    return self;
}

- (DATASource *)dataSource {
    if (!_dataSource) {
        NSFetchRequest *request = [[NSFetchRequest alloc] initWithEntityName:@"User"];
        request.sortDescriptors = @[[NSSortDescriptor sortDescriptorWithKey:@"firstLetterOfName" ascending:YES],
                                    [NSSortDescriptor sortDescriptorWithKey:@"name" ascending:YES]];

        _dataSource = [[DATASource alloc] initWithTableView:self.tableView
                                             cellIdentifier:CellIdentifier
                                               fetchRequest:request
                                                mainContext:self.dataStack.mainContext
                                                sectionName:@"firstLetterOfName"
                                              configuration:^(UITableViewCell * _Nonnull cell, NSManagedObject * _Nonnull item, NSIndexPath * _Nonnull indexPath) {
                                                  cell.textLabel.text = [item valueForKey:@"name"];
                                              }];
    }

    return _dataSource;
}

- (void)viewDidLoad {
    [super viewDidLoad];

    [self.tableView registerClass:[UITableViewCell class] forCellReuseIdentifier:CellIdentifier];

    self.navigationItem.rightBarButtonItem = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemAdd target:self action:@selector(addAction)];

    self.tableView.dataSource = self.dataSource;

    NSManagedObject *object = [self.dataSource objectAtIndexPath:[NSIndexPath indexPathForItem:0 inSection:0]];
    NSLog(@"object: %@", object);
}

- (void)addAction {
    [Helper addNewUserWithDataStack:self.dataStack];
}

@end

File: SwiftDemo/TableViewControllerWithSections/TableViewControllerWithSections.swift
import UIKit
import DATAStack
import CoreData

class TableViewControllerWithSections: UITableViewController {
    unowned let dataStack: DATAStack

    lazy var dataSource: DATASource = {
        let request: NSFetchRequest = NSFetchRequest<NSFetchRequestResult>(entityName: "User")
        request.sortDescriptors = [
            NSSortDescriptor(key: "firstLetterOfName", ascending: true),
            NSSortDescriptor(key: "name", ascending: true),
        ]

        let dataSource = DATASource(tableView: self.tableView, cellIdentifier: "Cell", fetchRequest: request, mainContext: self.dataStack.mainContext, sectionName: "firstLetterOfName")
        dataSource.delegate = self

        return dataSource
    }()

    init(dataStack: DATAStack) {
        self.dataStack = dataStack

        super.init(style: .plain)
    }

    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    override func viewDidLoad() {
        super.viewDidLoad()

        self.tableView.register(UITableViewCell.self, forCellReuseIdentifier: "Cell")
        self.navigationItem.rightBarButtonItem = UIBarButtonItem(barButtonSystemItem: .add, target: self, action: #selector(saveAction))
        self.tableView.dataSource = self.dataSource
    }

    @objc func saveAction() {
        Helper.addNewUser(dataStack: self.dataStack)
    }

    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        if let user = dataSource.object(indexPath) {
            let name = user.value(forKey: "name") as! String + "1"
            user.setValue(name, forKey: "name")
            try! dataStack.mainContext.save()
        }
    }
}

extension TableViewControllerWithSections: DATASourceDelegate {

    func dataSource(_ dataSource: DATASource, configureTableViewCell cell: UITableViewCell, withItem item: NSManagedObject, atIndexPath indexPath: IndexPath) {
        cell.textLabel?.text = item.value(forKey: "name") as? String ?? ""
    }
}
